routing/index_graph_starter.cpp

```diff
@@ -1,180 +1,396 @@
+Segment GetReverseSegment(Segment const & segment)
+{
+  return Segment(segment.GetMwmId(), segment.GetFeatureId(), segment.GetSegmentIdx(),
+                 !segment.IsForward());
+}
+
+Junction CalcProjectionToSegment(Segment const & segment, m2::PointD const & point,
                                  WorldGraph & graph)
 {
+  auto const & begin = graph.GetJunction(segment, false /* front */);
+  auto const & end = graph.GetJunction(segment, true /* front */);
   m2::ProjectionToSection<m2::PointD> projection;
-  projection.SetBounds(graph.GetPoint(segment, false /* front */),
-                       graph.GetPoint(segment, true /* front */));
-  return Junction(projection(junction.GetPoint()), junction.GetAltitude());
+  projection.SetBounds(begin.GetPoint(), end.GetPoint());
+  auto projectedPoint = projection(point);
+  auto const distBeginToEnd = graph.GetEstimator().CalcLeapWeight(begin.GetPoint(), end.GetPoint());
+
+  double constexpr kEpsMeters = 2.0;
+  if (my::AlmostEqualAbs(distBeginToEnd, 0.0, kEpsMeters))
+    return Junction(projectedPoint, begin.GetAltitude());
+
+  auto const distBeginToProjection =
+      graph.GetEstimator().CalcLeapWeight(begin.GetPoint(), projectedPoint);
+  auto const altitude = begin.GetAltitude() + (end.GetAltitude() - begin.GetAltitude()) *
+                                                  distBeginToProjection / distBeginToEnd;
+  return Junction(projectedPoint, altitude);
 }
 }  // namespace
 
 namespace routing
 {
-// static
-Segment constexpr IndexGraphStarter::kStartFakeSegment;
-Segment constexpr IndexGraphStarter::kFinishFakeSegment;
+void IndexGraphStarter::FakeGraph::AddFakeVertex(Segment const & existentSegment,
+                                                 Segment const & newSegment,
+                                                 FakeVertex const & newVertex, bool isOutgoing,
+                                                 bool isPartOfReal, Segment const & realSegment)
+{
+  auto const & segmentFrom = isOutgoing ? existentSegment : newSegment;
+  auto const & segmentTo = isOutgoing ? newSegment : existentSegment;
+  m_outgoing[segmentFrom].insert(segmentTo);
+  m_ingoing[segmentTo].insert(segmentFrom);
+  m_segmentToVertex[newSegment] = newVertex;
+  if (isPartOfReal)
+  {
+    m_realToFake[realSegment].insert(newSegment);
+    m_fakeToReal[newSegment] = realSegment;
+  }
+}
+
+Segment IndexGraphStarter::FakeGraph::GetSegment(FakeVertex const & vertex,
+                                                 uint32_t & newNumber) const
+{
+  for (auto const & kv : m_segmentToVertex)
+  {
+    if (kv.second == vertex)
+      return kv.first;
+  }
+
+  return GetFakeSegment(newNumber++);
+}
+
+void IndexGraphStarter::AddEnding(FakeEnding const & thisEnding, FakeEnding const & otherEnding,
+                                  bool isStart, bool strictForward, uint32_t & fakeNumerationStart)
+{
+  Segment const dummy = Segment();
```

Comment:
Why not to:
```c++
Segment const dummy;
```
?

Reply:
Linux build fails

Reply:
Could you please show me the compiler error message?

Reply:
Sure. "error: default initialization of an object of const type 'const routing::Segment' without a user-provided default constructor". http://angola.mapsme.mail.msk:8080/job/GitHubWatchdog/1864/BUILD_MODE=debug,label=Linux/console

We have
`Segment() = default;` in Segment.

Reply:
Okay.